atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/locks/Lockable.java

```diff
@@ -0,0 +1,133 @@
+    // negatives. It's allowed that something local is working on this item and Timelock has released the lock due to
+    // expiry and this is still True (since that still lets us minimise conflicts by not working on it locally),
+    // but it should not be true if something local has stopped working on this item and closed the inner class
+    private final AtomicBoolean isLocked = new AtomicBoolean(false);
+
+    private Lockable(
+            T inner, LockDescriptor lockDescriptor, TimelockService timeLock, Refreshable<Duration> lockTimeout) {
+        this.timeLock = timeLock;
+        this.inner = inner;
+        this.lockTimeout = lockTimeout;
+        this.lockDescriptor = lockDescriptor;
+    }
+
+    public static <T> Lockable<T> create(
+            T inner, LockDescriptor lockDescriptor, TimelockService timeLock, Refreshable<Duration> lockTimeout) {
+        return new Lockable<>(inner, lockDescriptor, timeLock, lockTimeout);
+    }
+
+    // For code owners:
+    // Do not use this for multiple locks!! If you're not careful, you can make it really easy to deadlock
+    // If you end up using this class and have a need for batch locks, implement a separate method that uses
+    // Timelocks batch locking methods and ensure there's a consistent ordering when taking out local locks
+    // to avoid circular waiting.
+    //    @MustBeClosed TODO: How do I get this working?? I'll track looking into this for another time.
+    @CheckReturnValue
+    public Optional<Inner> tryLock() {
+        if (isLocked.compareAndSet(false, true)) {
+            // We do not want the timeout to be too low to avoid a race condition where we give up too soon
+            LockRequest request = LockRequest.of(
+                    ImmutableSet.of(lockDescriptor), lockTimeout.get().toMillis());
+
+            try {
+                return timeLock.lock(request).getTokenOrEmpty().map(lockToken -> new Inner(lockToken));
+            } catch (Exception e) {
+                log.error("Failed to acquire lock", UnsafeArg.of("lockDescriptor", lockDescriptor), e);
+                isLocked.set(false);
+                return Optional.empty();
+            }
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    // Expected to be used in a try closeable.
+    public final class Inner implements AutoCloseable {
+        private final LockToken lockToken;
+
+        private Inner(LockToken lockToken) {
+            this.lockToken = lockToken;
+        }
+
+        public T getInner() {
+            return inner;
+        }
+
+        public Lockable<T> getOuter() {
+            return Lockable.this;
+        }
+
+        @Override
+        public void close() {
+            try {
+                timeLock.unlock(ImmutableSet.of(lockToken));
+            } finally {
+                isLocked.set(false); // We end up holding the local lock ever so slightly longer than necessary.
+            }
+        }
```

Comment:
I see the motivation for using a non-static inner class, though I'm not sure we should use it - we don't use this pattern much elsewhere in the codebase. Should we just have something that has a pair of `T` and a dispose/close method? That said you might have plans for the `getOuter()` method which I don't know at this stage

Reply:
Ah, I had a non-static inner class when I actually did need one for some reason when prototyping, then didn't revisit the decision. I'm not tied to it, although I believe I want inner to at least hold a reference to the lockable class so it doesn't get GC'd away when we're still using the lock!